app/models/pipeline_run.rb

```diff
@@ -153,38 +243,311 @@ def retry
+      "#{postprocess_output_s3_path}/#{TAXID_BYTERANGE_JSON_NAME}"
+    end
+  end
+
+  def output_ready?(output)
+    file_generated_since_run(s3_file_for(output))
+  end
+
+  def output_state_hash
+    h = {}
+    output_states.each do |o|
+      h[o.output] = o.state
+    end
+    h
+  end
+
+  def status_display(h = output_state_hash, results_finalized_var = results_finalized)
+    # Status display for the frontend.
+    if [1, FINALIZED_SUCCESS, FINALIZED_FAIL].include?(results_finalized_var) # 1 is for status_display_pre_result_monitor
+      if [h["taxon_byteranges"], h["taxon_counts"]].all? { |s| s == STATUS_LOADED }
+        "COMPLETE"
+      elsif h["taxon_counts"] == STATUS_LOADED
+        "COMPLETE*"
+      else
+        "FAILED"
+      end
+    elsif h["taxon_counts"] == STATUS_LOADED
+      # alignment succeeded, postprocessing in progress
+      "POST PROCESSING"
+    elsif h["ercc_counts"] == STATUS_LOADED
+      # host-filtering succeeded, alignment in progress
+      "ALIGNMENT"
+    else
+      # host-filtering in progress
+      "HOST FILTERING"
+    end
+  end
+
+  def pre_result_monitor?
+    results_finalized.nil?
+  end
+
+  def status_display_pre_result_monitor(run_stages)
+    # TODO: remove the need for this function by migrating old runs
+    state_by_output = {}
+    old_loaders_by_output = { "db_load_host_filtering" => "ercc_counts",
+                              "db_load_alignment" => "taxon_counts",
+                              "db_load_postprocess" => "taxon_byteranges" }
+    run_stages.each do |rs|
+      state_by_output[old_loaders_by_output[rs.load_db_command_func]] = rs.job_status
+    end
+    status_display(state_by_output, finalized)
+  end
+
+  def status_display_pre_run_stages
+    # TODO: remove the need for this function by migrating old runs
+    if %w[CHECKED SUCCEEDED].include?(job_status)
+      'COMPLETE'
+    elsif %w[FAILED ERROR].include?(job_status)
+      'FAILED'
+    elsif %w[RUNNING LOADED].include?(job_status)
+      'IN PROGRESS'
+    else
+      'WAITING'
+    end
+  end
+
+  def check_and_enqueue(output_state)
+    # If the pipeline monitor tells us that no jobs are running anymore,
+    # yet outputs are not available, we need to draw the conclusion that
+    # those outputs should be marked as failed. Otherwise we will never
+    # stop checking for them.
+    # [ TODO: move the check on "finalized" (column managed by pipeline_monitor)
+    #   to an S3 interface in order to give us the option of running pipeline_monitor
+    #   in a new environment that result_monitor does not have access to.
+    # ]
+    output = output_state.output
+    state = output_state.state
+    if finalized? && !output_ready?(output)
+      output_state.update(state: STATUS_FAILED)
+      return
+    end
+    # Otherwise, if the output is ready and has not yet been loaded or enqueued
+    # for loading, enqueue a loader job.
+    if output_ready?(output) && state == STATUS_UNKNOWN
```

Comment:
This is doing a lot of unnecessary aws S3 ops to check if output is ready.    You only need to check if STATUS is UNKNOWN.   I recommend rewriting as follows.

```
output = output_state.output
state = output_state.state
return unless state == STATUS_UNKNOWN
if output_ready?(output)
    output_state.update(state: STATUS_LOADING_QUEUED)
    Resque.enqueue(ResultMonitorLoader, id, output)
else if finalized?
    output_state.update(state: STATUS_FAILED)
end
```

Reply:
The explicit `return unless state == STATUS_UNKNOWN` is also consistent with the state transition diagram.   You might mention this function implements the Resource Monitor's state transitions from that diagram.  The only state the Resource Monitor is permitted to transition out of is the UNKNOWN state.   Which that diagram calls POLLING.   How about we just call it POLLING instead of UNKNOWN.

Reply:
ok changed